request: static built-in pre-completed request objects

hzhou opened the pull request:
## Pull Request Description

The current light weight request objects uses the dynamic object handle memory system. We probably can make it work as the light weight requests are created during init-time only. However, it creates constant worry as we need ensure the reference counting, request pool, per-vci critical section, etc. still works with light weight requests, thus, posing maintenance burden.

This PR proposes to create built-in per-completed static request objects so that it does not interfere with handle memory  system and allow simple check to by-pass most complications.

## Expected Impact

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Remove xfail from the test suite when fixing a test
* [ ] Commits are self-contained and do not do two things at once
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Passes whitespace checkers
* [ ] Passes warning tests
* [ ] Passes all tests
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] You or your company has a signed contributor's agreement on file with Argonne
* [ ] For non-Argonne authors, request an explicit comment from your companies PR approval manager

hzhou commented:
test:mpich/warnings/quick

hzhou commented:
test:mpich/warnings/quick

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
test:mpich/warnings/quick

hzhou commented:
test:mpich/warnings/quick

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

raffenet commented on src/include/mpir_request.h:
```diff
@@ -354,23 +364,23 @@ static inline MPIR_Request *MPIR_Request_create(MPIR_Request_kind_t kind, int po
 {
-    MPIR_Request *req;
-
-#ifdef HAVE_DEBUGGER_SUPPORT
```
We can't remove this debugger support.

hzhou replied:
What does the debugger support need? A unique `MPIR_Request` pointer?

raffenet replied:
Yes, one that includes the matching information (rank, tag, comm) for the posted send operation.

raffenet replied:
Er, more than just send operations.

hzhou replied:
Just worked on PR #4557 for `debuginfo` build. Seems we only support `sendq`, i.e. `SEND` and `PREQUEST_SEND` kinds, right? With builtin handle, I think we could add code in `dbginit.c` to special case the lightweight request. That way, we could isolate the debugger mess inside `dbginit.c`. Anyway, that will be a TODO in another time.

raffenet replied:
Hmm, yes, that seems doable since the API already takes the matching information separately from the request ptr.

raffenet commented:
You are right about the reference counting being a point of contention. Could we retain it, however, for `--enable-g` configurations? Otherwise we lose the ability to detect when user programs leak handles.

hzhou commented:
> You are right about the reference counting being a point of contention. Could we retain it, however, for `--enable-g` configurations? Otherwise we lose the ability to detect when user programs leak handles.

Yes, I think it can be managed.

raffenet commented:
Specially, I believe the configuration that detects handle leaks is `--enable-g=handlealloc`.

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
> 
> 
> Specially, I believe the configuration that detects handle leaks is `--enable-g=handlealloc`.

Can we treat `--enable-g=handlealloc` the same way as `HAVE_DEBUGGER_SUPPORT`?

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
test:mpich/warnings/quick

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

raffenet commented:
> > Specially, I believe the configuration that detects handle leaks is `--enable-g=handlealloc`.
> 
> Can we treat `--enable-g=handlealloc` the same way as `HAVE_DEBUGGER_SUPPORT`?

Do you mean merge the two configurations?

hzhou commented:
> > > Specially, I believe the configuration that detects handle leaks is `--enable-g=handlealloc`.
> > 
> > 
> > Can we treat `--enable-g=handlealloc` the same way as `HAVE_DEBUGGER_SUPPORT`?
> 
> Do you mean merge the two configurations?

Just here, something like:
```
#if defined(HAVE_DEBUGGER_SUPPORT) || defined(DEBUG_HANDLEALLOC) /* something like this*/
...
#else 
...
#endif
```

raffenet commented:
> > > > Specially, I believe the configuration that detects handle leaks is `--enable-g=handlealloc`.
> > > 
> > > 
> > > Can we treat `--enable-g=handlealloc` the same way as `HAVE_DEBUGGER_SUPPORT`?
> > 
> > 
> > Do you mean merge the two configurations?
> 
> Just here, something like:
> 
> ```
> #if defined(HAVE_DEBUGGER_SUPPORT) || defined(DEBUG_HANDLEALLOC) /* something like this*/
>    /* create the actual request object */
> #else 
>    /* return the builtin */
> ...
> #endif
> ```

Looks good to me.

hzhou commented:
test:mpich/custom --enable-debuginfo
netmod: ch4:ofi

hzhou commented:
test:mpich/custom --enable-debuginfo
netmod: ch4:ofi

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

hzhou commented:
Rebased and added commit [https://github.com/pmodels/mpich/pull/4550/commits/264d15c0d24a7e84287ad1c2bccbdd11004df666] to bypass debugger tracking for builtin-requests. Builtin requests are immediately completed, so I am not sure how much value is to fake a request objects just to show up on debugger's send-queue. By simply bypass builtin requests, at least the debugger build will not break.

Motivation: in order to get pre-completed requests onto the debugger's queue, not only we need create actual request objects, we also need support all the complications such as request pool, and per-vci critical sections. We can't hide these complications, but have to expose the vci parameter and explicit save/unsafe versions of `MPIR_Request_create_complete`. Not only it generates code churn, but also confuses the code -- for example, one may wonder why do we need track safety and vci of builtin requests.
